lambdas/data-migration2/tests/test-granulesAndFiles.js

```diff
@@ -591,6 +599,90 @@ test.serial('migrateGranuleRecord updates an already migrated record if the upda
   t.deepEqual(record.updated_at, new Date(newerGranule.updatedAt));
 });
 
+test.serial('migrateGranuleRecord supports undefined values in dynamo and overwrites defined values in Postgres when re-migrating', async (t) => {
+  const {
+    knex,
+    granulePgModel,
+    testCollection,
+    testExecution,
+    testProvider,
+    testPdr,
+  } = t.context;
+
+  const testGranule = generateTestGranule({
+    collectionId: constructCollectionId(testCollection.name, testCollection.version),
+    queryFields: { foo: cryptoRandomString({ length: 8 }) },
+    execution: testExecution.url,
+    provider: testProvider.name,
+    pdrName: testPdr.name,
+    status: 'completed',
+    version: cryptoRandomString({ length: 3 }),
+    updatedAt: Date.now() - 1000,
+  });
+
+  await createRejectableTransaction(knex, (trx) => migrateGranuleRecord(testGranule, trx));
+
+  // get a list of nullable granule keys
+  const nonNullablefields = [
+    'granuleId',
+    'collectionId',
+    'createdAt',
+    'updatedAt',
+    'status',
+    'execution',
+  ];
```

Comment:
I wouldn't mind some feedback on this. It gets really tricky because we have Postgres <> API granule translations sprinkled all over the place.

I'm particularly interested in thoughts about `execution`. I think it's right to call this non-nullable. `execution` does not exist on the granule postgres record, it's on the executions table with a foreign key pointing to the granule. If you update a granule via this migration with in Dynamo `execution: undefined`, I don't think we'd expect anything to happen. 

That logic might not actually be what we want though. We might want execution: undefined to then remove the execution from the executions table. The same question could be asked of `provider` and `pdrName`, although those aren't populated by lookups on other tables.

It's a very similar question to what happens with `files` when we do a migration and the files in Dynamo are different than PG.

Reply:
Or I guess it's a JOIN table, not a foreign key directly from granule to executions. Same question applies.

Reply:
`We might want execution: undefined to then remove the execution from the executions table.`

We do not want that to occur on a granule migration.   The reason this sort of thing occurs with files is because files is a *json object* in the original record, not a linking value.  

This is, however, made more complex because in P2 there *is* a divergence between datastores, namely in that anything already in postgres may have a set of executions associated with the granule.  A migration overwrite in dynamo from a record without an execution should be rare, but if it's there it implies the most recent write of the granule *did not have one* for some reason.    In either case we can't assume anything about the rest of the execution history for the overwrite granule.    

All of that said, we're assuming in case of migration overwrite that we're generally updating to fix a data sync issue, not a situation where only dynamo has been being written to and somehow the granule in postgres is functionally a different granule - and that's the only case where you'd really want a deletion of all the execution records.

Reply:
Discussed in a slack call. tl;dr is that we don't want a Dynamo -> Postgres Granule migration to remove executions or records in the join table if we have a null execution in Dynamo.